test/diffbinder.cpp

```diff
@@ -0,0 +1,67 @@
+#include <iostream>
+#include <fstream>
+#include <stdio.h>
+#include <vector>
+#include <string.h>
+
+int skip(const std::string& s1)
+{
+    if (s1.length()==0) return 1;
+    if (s1.length()>1) 
+    {
+        size_t comment=s1.find_first_of("//");
+        if (comment==s1.find_first_not_of(" \t")&&comment!=std::string::npos) return 2; //The comments should be ignored
+    }
+    if (s1.find("Source list file")!=std::string::npos) return 3;       //Same as the list of source files
+    if (s1.find("Modules list file")!=std::string::npos) return 4;      //Same as the list of modules
+    if (s1.find("File:")!=std::string::npos) return 5;                  //Same as the full file paths.
+    if (s1.find("pybind11::handle cl_type = cl;")!=std::string::npos) return 6;  //This could be optimized in binder
+    if (s1.find("__cxx11")!=std::string::npos) return 7;                         //This looks like implementation dependent
+    if (s1.find("assign")!=std::string::npos) return 8;                          // The assing is binded differently for some reasons
+    return 0;
+}
+int COMPARE_ASCII_FILES(const std::string& f1,const std::string& f2)
+{
+    std::fstream file1(f1.c_str()), file2(f2.c_str());
+    std::string string1, string2;
+    int j1,j2;
+    j1 = 0;
+    j2 = 0;
+    if (!file1.is_open()|| !file2.is_open()) 
+    {
+            std::cout << "Cannot open one of files" << f1<<f2<<"\n";
+            return EXIT_FAILURE;
+    }
+    std::cout <<"Run comparison"<< "\n";
+    while((!file1.eof())&&(!file2.eof()))
+    {
+        for (;;) {
+            j1++;
+            if (!std::getline(file1,string1)) break;
+            if (skip(string1)==0) break;
+        }
+        for (;;) {
+            j2++;
+            if (!std::getline(file2,string2)) break;
+            if (skip(string2)==0) break;
+        }
+        if(string1.compare(string2) != 0)
+        {
+            std::cout << j1<<"/"<<j2 << "-th strings are not equal" << f1<<f2<<"\n";
+            std::cout << "   ->" << string1 << "<-\n";
+            std::cout << "   ->" << string2 << "<-\n";
+            return EXIT_FAILURE;
+        }
+    }
+    return EXIT_SUCCESS;
+}
+int main(int argc, char** argv)
+{
+    if (argc!=3&&argc!=4) {
+        std::cout<<"Wrong number of arguments"<<std::endl;
+        return EXIT_FAILURE;
+    }
+    int status=COMPARE_ASCII_FILES(argv[1],argv[2]);
+    if (argc==4)  { if (strcmp(argv[3],"--always-success")==0)  return EXIT_SUCCESS; }
+    return status;
+}
```

Comment:
@andriish i added only some comments for this file and will add more when/if we decide if using custom-made utility is a sound approach. Here is my thoughts on this:

-  I do agree with your point that using just plain `diff` might not be optimal due to potential use on non GNU systems
- using custom C++ code might have benefits if Python is not present on target system
- however since _we do need python to run tests_, - why should not we just use Python scripts instead? 
- using Python script will allow to unify self-test suite with these tests so same comparison logic is used

Could you please argue why using C++ instead of Python will be a better choice? Thanks,

Reply:
OK. That is simple. Once everything is working, the simplest set of tests to run is just the diff test "-DBINDER_TEST_PYTHON_VERSIONS=0".   These test don't need python at all. One can also run them on the systems w/o pybind11 headers.

Reply:
> "finding Python interpreter"
Some systems have python, some have python2, some python3.  
To have a python interpreter in a build system (cmake or autotools) in a portable way one has to find it and not expect 
that python/python2/python3/python.exe is in PATH.
That stuff is always ugly.

>  the regex version from GCC-4.8 shipped with CentOS-7
Yes, I know. Very sad story. But hopefully, in the end,  there will be no need for regexps (i.e. the whole function 'skip' will not be needed) and the outputs will be always the same.

>command-line-options parsing
That can be added, there are standard superportable solutions. But diffbinder is an internal utility.

> C++ solution is superior or because current implementation is already written?
As for me, it is not superior because the task is trivial. For the same reason, one needs very little time to implement it in any language, so "already exist" is not a strong argument. 
But the C++ implementation is much more portable and does not require any scripting in cmake.

Practically: one can add an alternative 'diffbinder.py' and remove the c++ version in another MR.

An esoteric option:  write the diff as cmake function. Regexps and portability included.

Reply:
ok, lets accept this now, and we can revisit this later when/if expanding of this utility became an issue.